Sources/GRPCInterceptors/Tracing/ClientOTelTracingInterceptor.swift

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2024, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+public import GRPCCore
+internal import Synchronization
+internal import Tracing
+
+/// A client interceptor that injects tracing information into the request.
+///
+/// The tracing information is taken from the current `ServiceContext`, and injected into the request's
+/// metadata. It will then be picked up by the server-side ``ServerTracingInterceptor``.
+///
+/// For more information, refer to the documentation for `swift-distributed-tracing`.
+public struct ClientOTelTracingInterceptor: ClientInterceptor {
+  private let injector: ClientRequestInjector
+  private let emitEventOnEachWrite: Bool
+  private var serverHostname: String
+  private var networkTransportMethod: String
+
+  /// Create a new instance of a ``ClientOTelTracingInterceptor``.
+  ///
+  /// - Parameters:
+  ///  - severHostname: The hostname of the RPC server. This will be the value for the `server.address` attribute in spans.
+  ///  - networkTransportMethod: The transport in use (e.g. "tcp", "udp"). This will be the value for the
+  ///  `network.transport` attribute in spans.
+  ///  - emitEventOnEachWrite: If `true`, each request part sent and response part received will be recorded as a separate
+  ///  event in a tracing span. Otherwise, only the request/response start and end will be recorded as events.
+  public init(
+    serverHostname: String,
+    networkTransportMethod: String,
+    emitEventOnEachWrite: Bool = false
```

Comment:
This uses "each write" but also applies to messages which are received. Should we call this `traceEachMessage`?

Separate question: is this the right default? I think that would be the least surprising behaviour.

Reply:
Yeah I agree the naming isn't great but I couldn't think of anything that wasn't awful 😄 I like your suggestion, I'll change it.